tests/test_combiners.py

```diff
@@ -0,0 +1,172 @@
+from apache_beam.testing.util import assert_that
+from pytest_lazyfixture import lazy_fixture
+
+from pangeo_forge_recipes.combiners import CombineNested, CombineXarraySchemas, NestDim
+from pangeo_forge_recipes.patterns import CombineOp, DimKey, FilePattern, Index
+
+# to make these tests more isolated, we create a schema fixture
+# from scratch
+
+
+@pytest.fixture
+def pipeline():
+    # Runtime type checking doesn't play well with our Combiner
+    # https://github.com/apache/beam/blob/3cddfaf58c69acc624dac16df10357a78ececf59/sdks/python/apache_beam/transforms/core.py#L2505-L2509
+    options = PipelineOptions(runtime_type_check=False)
+    with TestPipeline(options=options) as p:
+        yield p
+
+
+def _get_schema(path):
+    ds = xr.open_dataset(path)
+    return ds.to_dict(data=False)
+
+
+def _expected_schema(pattern):
+    expected_ds = xr.open_mfdataset(item[1] for item in pattern.items())
+    expected_schema = expected_ds.to_dict(data=False)
+    expected_schema["chunks"] = {
+        "time": {pos: v for pos, v in enumerate(expected_ds.chunks["time"])}
+    }
+    return expected_schema
+
+
+# TODO: make this fixture leaner; don't need netcdf3 and netcdf4
+@pytest.fixture(
+    scope="module",
+    params=[
+        lazy_fixture("netcdf_local_file_pattern_sequential"),
+    ],
+    ids=["sequential"],
+)
+def schema_pcoll_concat(request):
+    pattern = request.param
+    expected_schema = _expected_schema(pattern)
+    return (
+        pattern,
+        expected_schema,
+        beam.Create(((key, _get_schema(url)) for key, url in pattern.items())),
+    )
+
+
+@pytest.fixture(
+    scope="module",
+    params=[
+        lazy_fixture("netcdf_local_file_pattern_sequential_multivariable"),
+    ],
+    ids=["sequential_multivariable"],
+)
+def schema_pcoll_concat_merge(request):
+    pattern = request.param
+    expected_schema = _expected_schema(pattern)
+    return (
+        pattern,
+        expected_schema,
+        beam.Create(((key, _get_schema(url)) for key, url in pattern.items())),
+    )
+
+
+def _get_concat_dim(pattern):
+    cdims = pattern.concat_dims
+    assert len(cdims) == 1, "Only one concat_dim allowed for now"
+    return cdims[0]
+
+
+def _strip_keys(item):
+    return item[1]
+
+
+def has_correct_schema(expected_schema):
+    def _check_results(actual):
+        assert len(actual) == 1
+        schema = actual[0]
+        assert schema == expected_schema
+
+    return _check_results
+
+
+def test_CombineXarraySchemas_concat_1D(schema_pcoll_concat, pipeline):
+    pattern, expected_schema, pcoll = schema_pcoll_concat
+    concat_dim = _get_concat_dim(pattern)
+
+    with pipeline as p:
+        input = p | pcoll
+        output = input | beam.CombineGlobally(
+            CombineXarraySchemas(DimKey(name=concat_dim, operation=CombineOp.CONCAT))
+        )
+        assert_that(output, has_correct_schema(expected_schema))
+
+
+def test_NestDim(schema_pcoll_concat_merge, pipeline):
```

Comment:
I'd like to better understand the use cases related to nested dimensions. Can you explain how these might be found in the wild?

Reply:
Thanks, this helps me understand what's going on better. I think nesting is a fine terminology, though I'll let you know if I think of something better. I still am a bit fuzzy on all of the details, but it occurs to me: How about we have the transforms related to nesting be private? Since users will probably only make use of the results of some top level `GroupByKey` or combiner, we don't want to risk messing things up with internal-only transforms.